pocs/camera/libasi.py

```diff
@@ -64,17 +70,30 @@ def get_cameras(self):
         # Get the IDs
         cameras = {}
         for camera_index in range(n_cameras):
-            # Can get IDs without opening cameras by parsing the 'name' string, which has the format
-            # "model(string_id)" if the string ID has been set, otherwise it is just the model name.
             info = self.get_camera_property(camera_index)
-            model, _, string_id = info['name'].partition('(')
-            if not string_id:
-                self.logger.warning("Found ZWO ASI camera with no ID set")
-                break
-            assert string_id.endswith(')'), self.logger.error("Expected ID enclosed in ()")
-            string_id = string_id[:-1]
-            cameras[string_id] = info['camera_ID']
-
+            camera_ID = info['camera_ID']
+            self.open_camera(camera_ID)
+            try:
+                serial_number = self.get_serial_number(camera_ID)
+            except RuntimeError as err:
+                # If at first you don't succeed, try, except, else, finally again.
+                self.logger.error(f"Error getting serial number: {err}")
+                try:
+                    string_ID = self.get_ID(camera_ID)
+                except RuntimeError as err:
+                    self.logger.error(f"Error getting string ID: {err}")
+                    msg = f"Skipping ZWO ASI camera {camera_ID} with no serial number or string ID."
+                    self.logger.warning(msg)
```

Comment:
You have a logger `error` with no serial id but a `warning` here when it's neither. Seems like you would want the error here as that means the camera is being skipped, but just a warning on the lack of real serial id.

Reply:
Yeah, I kept changing my mind about whether these 3 should be `warning` or `error`. `warning`, `warning`, `error` is probably more logical & self-consistent than what I currently have.

Part of the problem is that `._call_function()` doesn't (currently) make any distinctions based on which error code it gets back from the driver library, it just raises a generic `RuntimeError` for everything and embeds the error name in the exception's sole argument. This makes it a bit awkward for code that's catching these exceptions to respond in the most appropriate way. What `._call_function()` should probably be doing is raising `RuntimeError` with multiple arguments e.g. (human readable summary, SDK function name, error code, error name) so that it's easier for higher level code to work out what has happened or, even better, `._call_function()` should be raising different custom exceptions for each type of error, again with multiple attributes to make it easy to write conditional error handling code.

As an example, `ASIGetSerialNumber()` can return `ASI_ERROR_GENERAL_ERROR` if the camera doesn't have a serial number (which probably should just be a `warning` if it happens in `get_cameras()`) or `ASI_ERROR_CAMERA_CLOSED` if the camera is not open (which would definitely merit an `error`).

Reply:
Written up as #905, will be the subject of another PR.